src/Access/KerberosInit.cpp

```diff
@@ -0,0 +1,213 @@
+    {
+        // Use an existing cache for the client principal if we can.
+        ret = krb5_cc_cache_match(k5.ctx, k5.me, &k5.out_cc);
+        if (ret && ret != KRB5_CC_NOTFOUND)
+            throw Exception("Error while searching for cache for " + principal, ErrorCodes::KERBEROS_ERROR);
+        if (!ret)
+        {
+            LOG_DEBUG(log,"Using default cache: {}", krb5_cc_get_name(k5.ctx, k5.out_cc));
+            k5.switch_to_cache = 1;
+        }
+        else if (defcache_princ != nullptr)
+        {
+            // Create a new cache to avoid overwriting the initialized default cache.
+            ret = krb5_cc_new_unique(k5.ctx, deftype, nullptr, &k5.out_cc);
+            if (ret)
+                throw Exception("Error while generating new cache", ErrorCodes::KERBEROS_ERROR);
+            LOG_DEBUG(log,"Using default cache: {}", krb5_cc_get_name(k5.ctx, k5.out_cc));
+            k5.switch_to_cache = 1;
+        }
+    }
+
+    // Use the default cache if we haven't picked one yet.
+    if (k5.out_cc == nullptr)
+    {
+        k5.out_cc = defcache;
+        defcache = nullptr;
+        LOG_DEBUG(log,"Using default cache: {}", krb5_cc_get_name(k5.ctx, k5.out_cc));
+    }
+
+    ret = krb5_unparse_name(k5.ctx, k5.me, &k5.name);
+    if (ret)
+        throw Exception("Error when unparsing name", ErrorCodes::KERBEROS_ERROR);
+    LOG_DEBUG(log,"Using principal: {}", k5.name);
+
+    memset(&my_creds, 0, sizeof(my_creds));
+    ret = krb5_get_init_creds_opt_alloc(k5.ctx, &options);
+    if (ret)
+        throw Exception("Error in options allocation", ErrorCodes::KERBEROS_ERROR);
+
+    // Resolve keytab
+    ret = krb5_kt_resolve(k5.ctx, keytab_file.c_str(), &keytab);
+    if (ret)
+        throw Exception("Error in resolving keytab "+keytab_file, ErrorCodes::KERBEROS_ERROR);
+    LOG_DEBUG(log,"Using keytab: {}", keytab_file);
+
+    if (k5.in_cc)
+    {
+        ret = krb5_get_init_creds_opt_set_in_ccache(k5.ctx, options, k5.in_cc);
+        if (ret)
+            throw Exception("Error in setting input credential cache", ErrorCodes::KERBEROS_ERROR);
+    }
+    ret = krb5_get_init_creds_opt_set_out_ccache(k5.ctx, options, k5.out_cc);
+    if (ret)
+        throw Exception("Error in setting output credential cache", ErrorCodes::KERBEROS_ERROR);
+
+    // Action: init or renew
+    LOG_DEBUG(log,"Trying to renew credentials");
+    ret = krb5_get_renewed_creds(k5.ctx, &my_creds, k5.me, k5.out_cc, nullptr);
+    if (ret)
+    {
+        LOG_DEBUG(log,"Renew failed ({}). Trying to get initial credentials", ret);
+        ret = krb5_get_init_creds_keytab(k5.ctx, &my_creds, k5.me, keytab, 0, nullptr, options);
+        if (ret)
+            throw Exception("Error in getting initial credentials", ErrorCodes::KERBEROS_ERROR);
+        else
+            LOG_DEBUG(log,"Got initial credentials");
+    }
+    else
+    {
+        LOG_DEBUG(log,"Successful renewal");
+        ret = krb5_cc_initialize(k5.ctx, k5.out_cc, k5.me);
```

Comment:
In case of renewal action we need to clean the existing cache by krb5_cc_initialize and then save new creditials into the cache by krb5_cc_store_cred. That’s if how we update for example, ticket’s expiration date, for example.